### ISSUE TITLE
leaking fds in VNCServer mode until explodes from out of fds

### ISSUE THREAD

**Issue by twhitehead** [CanonicalLtd/lightdm#19]
We are running lightdm 1.25.0 (CentOS7) with

```
[VNCServer]
enabled=true
command=Xvnc -securitytypes vencrypt,x509none -x509cert ... -x509key ...
depth=24
```

and it seems that lightdm is not closing the VNC sockets.  That is, the total number of sockets just keeps increasing until it explodes with out of file descriptor messages.

Here is a listing of the _/proc/<pid>/fd_ directory after running for about 30-60min.

```
lr-x------ 1 root root 64 May 14 14:41 0 -> /dev/null
lrwx------ 1 root root 64 May 14 14:41 1 -> socket:[73145]
lr-x------ 1 root root 64 May 14 14:41 10 -> /var/lib/lightdm-data
lrwx------ 1 root root 64 May 14 15:04 100 -> socket:[67840]
lrwx------ 1 root root 64 May 14 15:04 101 -> socket:[67843]
lrwx------ 1 root root 64 May 14 15:04 102 -> socket:[67846]
lrwx------ 1 root root 64 May 14 15:04 103 -> socket:[59411]
lrwx------ 1 root root 64 May 14 15:04 104 -> socket:[29866]
lrwx------ 1 root root 64 May 14 15:04 105 -> socket:[37109]
lrwx------ 1 root root 64 May 14 15:04 106 -> socket:[46129]
lrwx------ 1 root root 64 May 14 15:04 107 -> socket:[10326]
lrwx------ 1 root root 64 May 14 15:04 108 -> socket:[10329]
lrwx------ 1 root root 64 May 14 15:04 109 -> socket:[22757]
lrwx------ 1 root root 64 May 14 14:41 11 -> socket:[34575]
lrwx------ 1 root root 64 May 14 15:04 110 -> socket:[26996]
lrwx------ 1 root root 64 May 14 15:04 111 -> socket:[70812]
lrwx------ 1 root root 64 May 14 15:04 112 -> socket:[37112]
lrwx------ 1 root root 64 May 14 15:04 113 -> socket:[10332]
lrwx------ 1 root root 64 May 14 15:04 114 -> socket:[22760]
lrwx------ 1 root root 64 May 14 15:04 115 -> socket:[18666]
lrwx------ 1 root root 64 May 14 15:04 116 -> socket:[10335]
lrwx------ 1 root root 64 May 14 15:04 117 -> socket:[59414]
lrwx------ 1 root root 64 May 14 15:04 118 -> socket:[29869]
lrwx------ 1 root root 64 May 14 15:04 119 -> socket:[29874]
lrwx------ 1 root root 64 May 14 14:41 12 -> socket:[34577]
lrwx------ 1 root root 64 May 14 15:04 120 -> socket:[19970]
lrwx------ 1 root root 64 May 14 15:04 121 -> socket:[29877]
lrwx------ 1 root root 64 May 14 15:04 122 -> socket:[29880]
lrwx------ 1 root root 64 May 14 15:04 123 -> socket:[10339]
lrwx------ 1 root root 64 May 14 14:41 13 -> socket:[70767]
lrwx------ 1 root root 64 May 14 14:41 14 -> socket:[67769]
lrwx------ 1 root root 64 May 14 14:41 15 -> socket:[68681]
lrwx------ 1 root root 64 May 14 14:41 16 -> socket:[67772]
lrwx------ 1 root root 64 May 14 14:41 17 -> socket:[29834]
lrwx------ 1 root root 64 May 14 14:41 18 -> socket:[36215]
lrwx------ 1 root root 64 May 14 14:41 19 -> socket:[22697]
lrwx------ 1 root root 64 May 14 14:41 2 -> socket:[73145]
lrwx------ 1 root root 64 May 14 14:41 20 -> socket:[22699]
lrwx------ 1 root root 64 May 14 14:41 21 -> socket:[26960]
lrwx------ 1 root root 64 May 14 14:41 22 -> socket:[59399]
lrwx------ 1 root root 64 May 14 14:41 23 -> socket:[29843]
lrwx------ 1 root root 64 May 14 14:41 24 -> socket:[68686]
lrwx------ 1 root root 64 May 14 14:41 25 -> socket:[46111]
lrwx------ 1 root root 64 May 14 14:41 26 -> socket:[66711]
lrwx------ 1 root root 64 May 14 14:41 27 -> socket:[56410]
l-wx------ 1 root root 64 May 14 14:41 28 -> pipe:[13640]
lrwx------ 1 root root 64 May 14 14:41 29 -> socket:[46108]
lrwx------ 1 root root 64 May 14 14:41 3 -> anon_inode:[eventfd]
lrwx------ 1 root root 64 May 14 14:41 30 -> socket:[67775]
lr-x------ 1 root root 64 May 14 14:41 31 -> pipe:[13641]
lr-x------ 1 root root 64 May 14 14:41 32 -> /ram/var/lib/sss/mc/passwd
lrwx------ 1 root root 64 May 14 14:41 33 -> socket:[68696]
lrwx------ 1 root root 64 May 14 14:41 34 -> socket:[56413]
lrwx------ 1 root root 64 May 14 14:41 35 -> socket:[70790]
lrwx------ 1 root root 64 May 14 14:41 36 -> socket:[59402]
lrwx------ 1 root root 64 May 14 14:41 37 -> socket:[26980]
lrwx------ 1 root root 64 May 14 14:41 38 -> socket:[46122]
lrwx------ 1 root root 64 May 14 14:41 39 -> socket:[29848]
lr-x------ 1 root root 64 May 14 14:41 4 -> pipe:[34569]
lrwx------ 1 root root 64 May 14 14:49 40 -> socket:[56416]
lrwx------ 1 root root 64 May 14 14:49 41 -> socket:[18645]
lrwx------ 1 root root 64 May 14 14:49 42 -> socket:[10296]
lrwx------ 1 root root 64 May 14 14:49 43 -> socket:[66714]
lrwx------ 1 root root 64 May 14 14:49 44 -> socket:[10299]
lrwx------ 1 root root 64 May 14 14:49 45 -> socket:[10302]
lrwx------ 1 root root 64 May 14 14:49 46 -> socket:[10305]
lrwx------ 1 root root 64 May 14 14:49 47 -> socket:[46123]
lrwx------ 1 root root 64 May 14 14:49 48 -> socket:[67796]
lrwx------ 1 root root 64 May 14 14:49 49 -> socket:[18648]
l-wx------ 1 root root 64 May 14 14:41 5 -> pipe:[34569]
lrwx------ 1 root root 64 May 14 14:49 50 -> socket:[67799]
lrwx------ 1 root root 64 May 14 14:49 51 -> socket:[29851]
lrwx------ 1 root root 64 May 14 14:49 52 -> socket:[67802]
lrwx------ 1 root root 64 May 14 14:50 53 -> socket:[68699]
lrwx------ 1 root root 64 May 14 14:50 54 -> socket:[18651]
lrwx------ 1 root root 64 May 14 14:50 55 -> socket:[67805]
lrwx------ 1 root root 64 May 14 14:50 56 -> socket:[66717]
lrwx------ 1 root root 64 May 14 14:50 57 -> socket:[66720]
lrwx------ 1 root root 64 May 14 14:50 58 -> socket:[10308]
lrwx------ 1 root root 64 May 14 14:53 59 -> socket:[28727]
l-wx------ 1 root root 64 May 14 14:41 6 -> /local/var/log/lightdm/lightdm.log
lrwx------ 1 root root 64 May 14 14:53 60 -> socket:[67809]
lrwx------ 1 root root 64 May 14 14:53 61 -> socket:[28730]
lrwx------ 1 root root 64 May 14 14:53 62 -> socket:[59405]
lrwx------ 1 root root 64 May 14 14:53 63 -> socket:[70797]
lrwx------ 1 root root 64 May 14 14:53 64 -> socket:[68702]
lrwx------ 1 root root 64 May 14 14:53 65 -> socket:[66723]
lrwx------ 1 root root 64 May 14 14:53 66 -> socket:[68705]
lrwx------ 1 root root 64 May 14 14:53 67 -> socket:[56425]
lrwx------ 1 root root 64 May 14 14:53 68 -> socket:[37071]
lrwx------ 1 root root 64 May 14 14:53 69 -> socket:[37074]
lrwx------ 1 root root 64 May 14 14:41 7 -> anon_inode:[eventfd]
lrwx------ 1 root root 64 May 14 14:53 70 -> socket:[46126]
lrwx------ 1 root root 64 May 14 14:53 71 -> socket:[29854]
lrwx------ 1 root root 64 May 14 14:53 72 -> socket:[67812]
lrwx------ 1 root root 64 May 14 14:53 73 -> socket:[18654]
lrwx------ 1 root root 64 May 14 14:55 74 -> socket:[18657]
lrwx------ 1 root root 64 May 14 14:55 75 -> socket:[67825]
lrwx------ 1 root root 64 May 14 14:55 76 -> socket:[67828]
lrwx------ 1 root root 64 May 14 14:55 77 -> socket:[10311]
lrwx------ 1 root root 64 May 14 14:55 78 -> socket:[29857]
lrwx------ 1 root root 64 May 14 14:55 79 -> socket:[26990]
lrwx------ 1 root root 64 May 14 14:41 8 -> anon_inode:[eventfd]
lrwx------ 1 root root 64 May 14 15:04 80 -> socket:[18660]
lrwx------ 1 root root 64 May 14 15:04 81 -> socket:[18663]
lrwx------ 1 root root 64 May 14 15:04 82 -> socket:[22751]
lrwx------ 1 root root 64 May 14 15:04 83 -> socket:[59408]
lrwx------ 1 root root 64 May 14 15:04 84 -> socket:[29860]
lrwx------ 1 root root 64 May 14 15:04 85 -> socket:[22754]
lrwx------ 1 root root 64 May 14 15:04 86 -> socket:[10314]
lrwx------ 1 root root 64 May 14 15:04 87 -> socket:[29863]
lrwx------ 1 root root 64 May 14 15:04 88 -> socket:[37100]
lrwx------ 1 root root 64 May 14 15:04 89 -> socket:[19941]
lrwx------ 1 root root 64 May 14 14:41 9 -> socket:[74967]
lrwx------ 1 root root 64 May 14 15:04 90 -> socket:[28739]
lrwx------ 1 root root 64 May 14 15:04 91 -> socket:[67831]
lrwx------ 1 root root 64 May 14 15:04 92 -> socket:[67834]
lrwx------ 1 root root 64 May 14 15:04 93 -> socket:[67837]
lrwx------ 1 root root 64 May 14 15:04 94 -> socket:[70809]
lrwx------ 1 root root 64 May 14 15:04 95 -> socket:[37103]
lrwx------ 1 root root 64 May 14 15:04 96 -> socket:[10317]
lrwx------ 1 root root 64 May 14 15:04 97 -> socket:[10320]
lrwx------ 1 root root 64 May 14 15:04 98 -> socket:[26993]
lrwx------ 1 root root 64 May 14 15:04 99 -> socket:[10323]
```

I realize it would be better if I tested the latest version and such, but I didn't find any new commits mentioning file descriptors since the `1.25.0` tag, so figured I would mention it here first in case it is something easy for someone familiar with the code to address.

Thanks!  -Tyson

**Comment by twhitehead** [comment]
Here is a log from running with `G_MESSAGES_DEBUG=all`

```
** (lightdm:22846): DEBUG: Logging to /var/log/lightdm/lightdm.log
** (lightdm:22846): DEBUG: Starting Light Display Manager 1.25.0, UID=0 PID=22846
** (lightdm:22846): DEBUG: Loading configuration dirs from /usr/share/lightdm/lightdm.conf.d
** (lightdm:22846): DEBUG: Loading configuration from /usr/share/lightdm/lightdm.conf.d/50-backup-logs.conf
** (lightdm:22846): DEBUG: Loading configuration from /usr/share/lightdm/lightdm.conf.d/50-disable-guest.conf
** (lightdm:22846): DEBUG: Loading configuration from /usr/share/lightdm/lightdm.conf.d/50-minimum-vt.conf
** (lightdm:22846): DEBUG: Loading configuration from /usr/share/lightdm/lightdm.conf.d/50-session-wrapper.conf
** (lightdm:22846): DEBUG: Loading configuration from /usr/share/lightdm/lightdm.conf.d/50-user-authority-in-system-dir.conf
** (lightdm:22846): DEBUG: Loading configuration from /usr/share/lightdm/lightdm.conf.d/50-xserver-command.conf
** (lightdm:22846): DEBUG: Loading configuration from /usr/share/lightdm/lightdm.conf.d/90-slick-greeter.conf
** (lightdm:22846): DEBUG: Loading configuration dirs from /usr/local/share/lightdm/lightdm.conf.d
** (lightdm:22846): DEBUG: Loading configuration dirs from /etc/xdg/lightdm/lightdm.conf.d
** (lightdm:22846): DEBUG: Loading configuration from /etc/lightdm/lightdm.conf
** (lightdm:22846): DEBUG: Registered seat module local
** (lightdm:22846): DEBUG: Registered seat module xremote
** (lightdm:22846): DEBUG: Registered seat module unity
** (lightdm:22846): DEBUG: Using D-Bus name org.freedesktop.DisplayManager
** (lightdm:22846): DEBUG: Monitoring logind for seats
** (lightdm:22846): DEBUG: Acquired bus name org.freedesktop.DisplayManager
** (lightdm:22846): DEBUG: Starting VNC server on TCP/IP port 5900

** (lightdm:22846): WARNING **: Failed to create IPv6 VNC socket: Error binding to address: Address already in use
** (lightdm:22846): DEBUG: Loading users from org.freedesktop.Accounts
```

At this point, if I do an ls on fds, I get

```
lrwx------ 1 root root 64 May 14 15:59 0 -> /dev/pts/1
lrwx------ 1 root root 64 May 14 15:59 1 -> /dev/pts/1
lrwx------ 1 root root 64 May 14 15:59 10 -> anon_inode:[eventfd]
lr-x------ 1 root root 64 May 14 15:59 11 -> /var/lib/lightdm-data
lrwx------ 1 root root 64 May 14 15:59 2 -> /dev/pts/1
lrwx------ 1 root root 64 May 14 15:59 3 -> anon_inode:[eventfd]
lr-x------ 1 root root 64 May 14 15:59 4 -> pipe:[83117]
l-wx------ 1 root root 64 May 14 15:59 5 -> pipe:[83117]
l-wx------ 1 root root 64 May 14 15:59 6 -> /local/var/log/lightdm/lightdm.log
lrwx------ 1 root root 64 May 14 15:59 7 -> anon_inode:[eventfd]
lrwx------ 1 root root 64 May 14 15:59 8 -> socket:[83121]
lrwx------ 1 root root 64 May 14 15:59 9 -> socket:[82674]
```

Wait a few seconds with an open port, and a random computer on the internet connects

```
** (lightdm:22846): DEBUG: Got VNC connection from 77.72.83.246:4476
** (lightdm:22846): DEBUG: Seat (null): Loading properties from config section Seat:*
** (lightdm:22846): DEBUG: Seat (null): Starting
** (lightdm:22846): DEBUG: Seat (null): Creating greeter session
** (lightdm:22846): DEBUG: Seat (null): Creating display server of type x
** (lightdm:22846): DEBUG: XServer 0: Logging to /var/log/lightdm/x-0.log
** (lightdm:22846): DEBUG: XServer 0: Writing X server authority to /var/run/lightdm/root/:0
** (lightdm:22846): DEBUG: XServer 0: Launching X Server
** (lightdm:22846): DEBUG: Launching process 22862: /usr/bin/Xvnc -securitytypes vencrypt,x509none -x509cert /etc/pki/tls/private/user-sharcnet-wildcard.pem -x509key /etc/pki/tls/private/user-sharcnet-wildcard.pem :0 -auth /var/run/lightdm/root/:0 -nolisten tcp -background none -inetd -geometry 1024x768 -depth 24
** (lightdm:22846): DEBUG: XServer 0: Waiting for ready signal from X server :0
** (lightdm:22846): DEBUG: Registering seat with bus path /org/freedesktop/DisplayManager/Seat0
** (lightdm:22846): DEBUG: Got signal 10 from process 22862
** (lightdm:22846): DEBUG: XServer 0: Got signal from X server :0
** (lightdm:22846): DEBUG: XServer 0: Connecting to XServer :0
** (lightdm:22846): DEBUG: XServer 0: Error connecting to XServer :0
** (lightdm:22846): DEBUG: Process 22862 exited with return value 0
** (lightdm:22846): DEBUG: XServer 0: X server stopped
** (lightdm:22846): DEBUG: XServer 0: Removing X server authority /var/run/lightdm/root/:0
** (lightdm:22846): DEBUG: Seat (null): Display server stopped
** (lightdm:22846): DEBUG: Seat (null): Stopping session
** (lightdm:22846): DEBUG: Seat (null): Session stopped
** (lightdm:22846): DEBUG: Seat (null): Stopping display server, no sessions require it
** (lightdm:22846): DEBUG: Seat (null): Stopping; greeter display server failed to start
** (lightdm:22846): DEBUG: Seat (null): Stopping
** (lightdm:22846): DEBUG: Seat (null): Stopped
```

Now if we list the fds, you can see that the associated socket (12 in this case) was never released

```
lrwx------ 1 root root 64 May 14 15:59 0 -> /dev/pts/1
lrwx------ 1 root root 64 May 14 15:59 1 -> /dev/pts/1
lrwx------ 1 root root 64 May 14 15:59 10 -> anon_inode:[eventfd]
lr-x------ 1 root root 64 May 14 15:59 11 -> /var/lib/lightdm-data
lrwx------ 1 root root 64 May 14 15:59 12 -> socket:[18760]
lrwx------ 1 root root 64 May 14 15:59 2 -> /dev/pts/1
lrwx------ 1 root root 64 May 14 15:59 3 -> anon_inode:[eventfd]
lr-x------ 1 root root 64 May 14 15:59 4 -> pipe:[83117]
l-wx------ 1 root root 64 May 14 15:59 5 -> pipe:[83117]
l-wx------ 1 root root 64 May 14 15:59 6 -> /local/var/log/lightdm/lightdm.log
lrwx------ 1 root root 64 May 14 15:59 7 -> anon_inode:[eventfd]
lrwx------ 1 root root 64 May 14 15:59 8 -> socket:[83121]
```

And so on it continues (the fd count goes up by one each time another computer connects) until you run out of fds and it dies badly.  Here is the CTRL+C killing it for completion

```
** (lightdm:22846): DEBUG: Got signal 2 from process 0
** (lightdm:22846): DEBUG: Caught Interrupt signal, shutting down
** (lightdm:22846): DEBUG: Stopping display manager
** (lightdm:22846): DEBUG: Display manager stopped
** (lightdm:22846): DEBUG: Stopping daemon
** (lightdm:22846): DEBUG: Exiting with return value 0
```

**Comment by twhitehead** [comment]
Tracked it down with a debugger only to find it has already been fixed with commit 8283fc7.
